---
title: "Lab Task 7"
author: "Group 6"
date: Sys.Date()
output:
  html_document:
    code_folding: show
editor_options: 
  markdown: 
    wrap: sentence
---

Group members:

- Shawn Yakir , 315714980,  shawn.yakir@mail.huji.ac.il
- Sarah Levitz, 324673623, sarah.levitz@mail.huji.ac.il
- Carmel Baris, 318455276, carmel.baris@mail.huji.ac.il 

### Clean the environment
```{r clean}
rm(list = ls())
```

### Libraries used

```{r libraries}
library('splines')
library('scales')
```


### Paths
```{r paths}
path = "C:/Users/user/Documents/Carmel/Projects/HUJI/Archive_HUJI/data-analysis-genomics/data"
# path = "/Users/alevi/Downloads/"
# path = "\\Users\\shawn\\Desktop\\"
```


### Helper functions
```{r helpers}
helpers_file = file.path(path, "help_funcs.R")
source(helpers_file)
helpers = list(loadRData = loadRData, binReads = binReads)
```


### Load data
```{r dat}
load(sprintf("%s/reads_100_A2.rda",path))
load(sprintf("%s/reads_100_B2.rda",path))
load(sprintf("%s/GC_100.rda",path))
GC_100=GC_100[1: (length(reads_100_A2)) ]
```

### Binning
```{r binning}
binsize = 25
chr2t_mat = matrix(reads_100_A2, nr = binsize)
reads2.5K_t2 = colSums(chr2t_mat)

chr2n_mat = matrix(reads_100_B2, nr = binsize)
reads2.5K_n2 = colSums(chr2n_mat)

gc_mat = matrix(GC_100, nr=binsize)
GC2.5K = colMeans(gc_mat)
```


### Outliers
```{r outliers}
# outliers of gc percent per bin
zeros_GC = GC2.5K == 0 #bins w/o GC content
# small_GC = GC2.5K <= 0.3 & GC2.5K > 0 #low-GC content bins (<30%)
# valid_GC = !(zeros_GC | small_GC) #bins w/ at least 30% GC content

# identify outliers of reads count per bin
z_scores_n2 = (reads2.5K_n2 - mean(reads2.5K_n2)) / sd(reads2.5K_n2)
z_scores_t2 = (reads2.5K_t2 - mean(reads2.5K_t2)) / sd(reads2.5K_t2)
xtreme_reads_n2 = abs(z_scores_n2) > 3
xtreme_reads_t2 = abs(z_scores_t2) > 3

# create new variables without outliers
outliers = (zeros_GC | xtreme_reads_t2 | xtreme_reads_n2)
# outliers = (outliers | (reads2.5K_t2<0.05) | (reads2.5K_n2<0.05))
reads2.5K_n2_clean = reads2.5K_n2[!outliers]
reads2.5K_t2_clean = reads2.5K_t2[!outliers]
GC2.5K_clean = GC2.5K[!outliers]
```




```{r normalizing}
# Use the rescale function from the scales package
reads2.5K_n2_normalized = scales::rescale(reads2.5K_n2_clean, to = c(0, 100))
reads2.5K_t2_normalized = scales::rescale(reads2.5K_t2_clean, to = c(0, 100))
```

```{r}
# Plot the normalized data
set.seed(50)
samp1 = sample(length(reads2.5K_n2), 10000)
# Plot adjacent
par(mfcol = c(2, 1), mar = c(1, 5, 1, 1))
plot(GC2.5K_clean[samp1], reads2.5K_n2_normalized[samp1], cex = 0.4, pch = 20, ylab = "Healthy 2", ylim = c(1.5, 35), xlim = c(0.2, 0.75))
plot(GC2.5K_clean[samp1], reads2.5K_t2_normalized[samp1], cex = 0.4, pch = 20, ylab = "Tumor 2", ylim = c(0, 35), xlim = c(0.2, 0.75))
```

### Fitting bspline regression
```{r fit}
f = function(X, Y) {
  lm(subset = Y>1.5&Y<35,formula = Y ~ bs(X, degree = 3, knots = c(0.415, 0.45, 0.54), Boundary.knots = c(min(X), max(X))))
}

```

## Learn parameters from chromosome 1

```{r}
l_2n = length(reads2.5K_n2_normalized)

GC = GC2.5K_clean[1:l_2n]
reads_n2 = reads2.5K_n2_normalized[1:l_2n]
reads_t2 = reads2.5K_t2_normalized[1:l_2n]

mod_2n = f(GC,reads_n2)
mod_2t = f(GC,reads_t2)

pred_range = seq(0.2,0.75,0.01)
resp_n = predict(mod_2n,list(GC=pred_range))
preds2n = predict(mod_2n,list(GC=GC2.5K_clean))

par(mfcol = c(1,1),mar = c(3,5,3,3))
plot(GC2.5K_clean[samp1],reads2.5K_n2_normalized[samp1],ylim = c(0,35),xlim = c(0.2,.75), cex = 0.5, pch=20, main = "Healthy 2")
lines(pred_range, resp_n,col = 4,lw=3)

```


```{r}
GC=GC2.5K_clean[1:l_2t]
mod_2t = lm(reads2.5K_t2_normalized~bs(GC, degree = 3),subset = (reads2.5K_t2_normalized>1.5)&(reads2.5K_t2_normalized<40))

pred_range = seq(0.2,0.75,0.01)
resp_t = predict(mod_2t,list(GC=pred_range))
preds2t = predict(mod_2t,list(GC=GC2.5K_clean))

par(mfcol = c(1,1),mar = c(3,3,3,3))
plot(GC2.5K_clean[samp1],reads2.5K_t2_normalized[samp1],ylim = c(0,35),xlim = c(0.2,.75), cex = 0.5, pch=20, main = "Tumor 2")
lines(pred_range, resp_t,col = 2,lw=3)
lines(pred_range, resp_n,col = 4,lw=3)
```


## One sample corrections
```{r func_fix}
one_sample_correct = function(samp, preds) {
  one_correct = (samp)/(preds)
  return(one_correct)
}
```

```{r med_fix}
corrected_2n = one_sample_correct(reads2.5K_n2_normalized,preds2n[1:l_2n])
uncorrected_2n = (reads2.5K_n2_normalized)/median(reads2.5K_n2_normalized) 

corrected_2t = one_sample_correct(reads2.5K_t2_normalized,preds2t[1:l_2t])
uncorrected_2t = (reads2.5K_t2_normalized)/median(reads2.5K_t2_normalized) 

```


```{r plot_tn_stacked}
range_a = 5000:6000

par(mfcol = c(4,1),mar = c(2,5,2,2))

plot(uncorrected_2n[range_a], ylim = c(0,2.5), cex = 0.5, pch=20, ylab = "Healthy 2", main = "Before Median Fix")
abline(h=1,col=4)
plot(uncorrected_2t[range_a], ylim = c(0,2.5), cex = 0.5, pch=20, ylab = "Tumor 2")
abline(h=1,col=4)

plot(corrected_2n[range_a], ylim = c(0,2.5), cex = 0.5, pch=20, ylab = "Healthy 2", main = "Corrected")
abline(h=1,col=4)
plot(corrected_2t[range_a], ylim = c(0,2.5), cex = 0.5, pch=20, ylab = "Tumor 2")
abline(h=1,col=4)

```

### Introduction 

This week we...

### Q1: 


