---
title: "Lab Task 8"
author: "Group 6"
date: "`r format(Sys.time(), '%d/%m/%y')`"
output:
  html_document:
    code_folding: show
editor_options: 
  markdown: 
    wrap: sentence
---

Group members:

- Shawn Yakir , 315714980,  shawn.yakir@mail.huji.ac.il
- Sarah Levitz, 324673623, sarah.levitz@mail.huji.ac.il
- Carmel Baris, 318455276, carmel.baris@mail.huji.ac.il 

### Clean the environment
```{r clean}
rm(list = ls())
```

### Libraries used

```{r libraries}
library('splines')
library('scales')
```


### Paths
```{r paths}
path = "C:/Users/user/Documents/Carmel/Projects/HUJI/Archive_HUJI/data-analysis-genomics/data"
# path = "/Users/alevi/Downloads/"
 # path = "\\Users\\shawn\\Desktop\\"
```


### Functions

Correcting functions:

```{r correcting_funcs}

### Median correction
fix_med = function(samp) {
  return(samp/median(samp,na.rm=TRUE))
}

### One sample correction
fix_gc_e = function(samp, preds, eps=0) {
  return((samp+eps)/(preds+eps))
}


### Two sample correction
fix_2samp_e = function(tumor, healthy, eps=0) {
  return((tumor+eps)/(healthy+eps))
}

```


Fitting bspline regression:

```{r fit}
f = function(X, Y, deg=3, knots=c(0.415, 0.45, 0.54), subs) {
  lm(Y ~ bs(X, degree = deg, knots = knots, Boundary.knots = c(min(X), max(X))),
     subset = subs)
}

```


### Load data
```{r dat}
load(sprintf("%s/reads_100_A2.rda",path))
load(sprintf("%s/reads_100_B2.rda",path))
load(sprintf("%s/GC_100.rda",path))
```

### Binning
```{r binning}
binsize = 25

chr2t_mat = matrix(reads_100_A2, nr = binsize)
chr2n_mat = matrix(reads_100_B2, nr = binsize)
GC_100=GC_100[1: (length(reads_100_A2)) ]
gc_mat = matrix(GC_100,nr=binsize)

reads_tumor = colSums(chr2t_mat)
reads_healthy = colSums(chr2n_mat)
GC = colMeans(gc_mat)

length(reads_healthy) == length(reads_tumor)
```


### Spatial Distribution of Healthy vs Tumor Reads
```{r plot_spatial}
# Plot the data
set.seed(50)
samp1 = sample(length(reads_healthy), 10000)
plot(GC[samp1], reads_tumor[samp1], cex = 0.4, pch = 20, ylab = "Tumor 2", ylim = c(0, 400), xlim = c(0.2, 0.75))
```


### Learn parameters from chromosome 1

```{r mod_pred}
l_2n = length(reads_healthy)
l_2t = length(reads_tumor)

myknots = c(0.415, 0.45, 0.54)
pred_range = seq(0.2,0.75,0.01)

mod_2n = lm(reads_healthy~bs(GC, degree = 3),subset = reads_healthy>50 & reads_healthy<350 )
mod_2t = lm(reads_tumor~bs(GC, degree = 3),subset = (reads_tumor>50)&(reads_tumor<400))

resp_n = predict(mod_2n,list(GC=pred_range))
preds2n = predict(mod_2n,list(GC=GC))

resp_t = predict(mod_2t,list(GC=pred_range))
preds2t = predict(mod_2t,list(GC=GC))

par(mfcol = c(1,1),mar = c(3,5,3,3))

plot(GC[samp1],reads_tumor[samp1],ylim = c(0,350),xlim = c(.2,.75), cex = 0.5, pch=20, main = "Tumor 2")
lines(pred_range, resp_t,col = 2,lw=3)

```



### Variance Decompisition

We'll start with a breakdown of the variance in the healthy sample after filtering out the outliers and after correcting the median.

```{r res_2n}
# Fit B-spline Model
l_2n = length(reads_healthy)
GC = GC[1:l_2n]
residuals_2n = reads_healthy - preds2n
```

```{r var_decompisition}
### Median Correction
fixed_2n = reads_healthy / median(reads_healthy)

### Variance Decomposition
total_var = var(reads_healthy)
poiss_var = mean(reads_healthy)/(median(reads_healthy)^2)
gc_variance = var(preds2n)
unexplained_variance = total_var-poiss_var - gc_variance

### Percentage of Variance
percentage_poisson = (poiss_var / total_var) * 100
percentage_gc = (gc_variance / total_var) * 100
percentage_unexplained = (unexplained_variance / total_var) * 100
```

```{r print_var_decompisition}
### Results
cat("Percentage of variance due to Poisson distribution:", round(percentage_poisson,13), "% < 0.001 %\n")
cat("Percentage of variance due to GC effect: ", round(percentage_gc,2), "%\n")
cat("Percentage of variance unexplained: ", round(percentage_unexplained,2), "%\n")

```

```{r med_fix}
fixed_2n_gc_med_eps = fix_gc_e(reads_healthy, preds2n[1:l_2n])
unfixed_2n_gc_med_eps = fix_med(reads_healthy)

fixed_2t_gc_med_eps = fix_gc_e (reads_tumor,preds2t[1:l_2t])
unfixed_2t_gc_med_eps = (reads_tumor)/median(reads_tumor) 

```


```{r plot_tn_stacked}
range_a = (25000000/2500):(30000000/2500)

par(mfcol = c(4,1),mar = c(2,5,2,2))

plot(unfixed_2n_gc_med_eps[range_a], ylim = c(0,2.5), cex = 0.5, pch=20, ylab = "Healthy 2", main = "Before Median Fix")
abline(h=1,col=4)
plot(unfixed_2t_gc_med_eps[range_a], ylim = c(0,2.5), cex = 0.5, pch=20, ylab = "Tumor 2")
abline(h=1,col=4)

plot(fixed_2n_gc_med_eps[range_a], ylim = c(0,2.5), cex = 0.5, pch=20, ylab = "Healthy 2", main = "After")
abline(h=1,col=4)
plot(fixed_2t_gc_med_eps[range_a], ylim = c(0,2.5), cex = 0.5, pch=20, ylab = "Tumor 2")
abline(h=1,col=4)

```




### Part 2 Q1: 

```{r fit_bs}
# Fit B-spline Regression Models
GC_healthy = GC
mod_healthy = lm(reads_healthy ~ bs(GC_healthy, degree = 3))
pred_healthy = predict(mod_healthy, list(GC_healthy = GC_healthy))

# Tumor Sample
reads_tumor = reads_tumor[1:l_2n]  # Ensure tumor data matches in length
mod_tumor = lm(reads_tumor ~ bs(GC_healthy, degree = 3))
pred_tumor = predict(mod_tumor, list(GC_healthy = GC_healthy))

```

```{r pred_range}
# Define the range for predictions
pred_range = seq(0.2, 0.75, 0.01)

# Predict values over the range for both models
preds_healthy = predict(mod_healthy, newdata = data.frame(GC_healthy = pred_range))
preds_tumor = predict(mod_tumor, newdata = data.frame(GC_healthy = pred_range))

```


```{r plot_combined}
# Combined Plot
par(mfcol = c(1, 1), mar = c(5, 5, 4, 2))

# Plot the data points first
plot(GC_healthy, reads_tumor, col = 'darkgray', pch = 20, cex = 0.5, ylab = "Normalized Reads", xlab = "GC Content", main = "Healthy vs Tumor Sample", ylim = c(0,400))
points(GC_healthy, reads_healthy, col = 'lightgray', pch = 20, cex = 0.5)

# Add the regression lines on top
lines(pred_range, preds_healthy, col = 'darkblue', lwd = 2)
lines(pred_range, preds_tumor, col = 'darkred', lwd = 2)

# Add legend
legend("topright", legend = c("Tumor Sample", "Healthy Sample"), col = c("darkgray", "lightgray"), pch = 20)

```


### Learn parameters from chromosome 1

```{r lengths}
l_2n = length(reads_healthy)
l_2t = length(reads_tumor)
GC=GC[1:l_2n]
length(GC)

```



### One sample corrections (aka only GC & median correction)
```{r one_samp_gc}
one_sample_gc_correct = function(samp, preds, eps) {
  one_correct = (samp+eps)/(preds+eps)
  adj_med = 1/median(one_correct)
  return(adj_med * one_correct)
}
```

```{r pmax}
# take maximum with 0
preds1n = pmax(predict(mod_2n,list(GC=GC)),0)
preds1t = pmax(predict(mod_2t,list(GC=GC)),0)


# One Sample with espilon, Normative
eps = 0.1

fixed_2n_gc_med_eps = one_sample_gc_correct(reads_healthy, preds1n[1:l_2n], eps)
unfixed_2n_gc_med_eps = one_sample_gc_correct(reads_healthy, rep(1,l_2n), eps)

# One Sample with espilon, Tumor
eps = 0.1
fixed_2t_gc_med_eps = one_sample_gc_correct (reads_tumor,preds1t[1:l_2t],eps)
unfixed_2t_gc_med_eps = one_sample_gc_correct(reads_tumor,rep(1,l_2t),eps)
```

```{r discontinuous_range}
par(mfcol = c(2,1),mar = c(0.5,5,2,2))
loc2bin = function(v, bin_size){
  return((v*1000000)/bin_size)}

bins = loc2bin(c(25,35,75,80),2500)
discontinuous_range = c(bins[1]:bins[2], bins[3]:bins[4])
# discontinuous_range = c(10000:14000,15000:16000)
```


```{r plot_gc_fix_tumor}
plot(unfixed_2t_gc_med_eps[discontinuous_range], 
     main = "GC Correction for Tumor Sample",
     ylab = "Before", xaxt = "n",
     ylim = c(0,2.5),cex = 0.2,pch=20, xlim = c(0,6000))
abline(h=1,col=4, lwd = 1.5)
abline(v=4000,col=6,lt = 2, lwd = 1.2)

# main_text = "GC Correction for Tumor Sample"
# sub_text = "One-Sample Correction for Tumor Sample"
# mtext(side=3, line=1, cex=1, main_text)
# mtext(side=1, line=1, cex=1, sub_text)

plot(fixed_2t_gc_med_eps[discontinuous_range], 
     ylab = "After", xaxt = "n",
     ylim = c(0,2.5),cex = 0.2,pch=20, xlim = c(0,6000))
abline(h=1,col=4, lwd = 1.5)
abline(v=4000,col=6,lt = 2, lwd = 1.2)

plot(unfixed_2n_gc_med_eps[discontinuous_range], ylab = "Before", xaxt = "n",
     main = "GC Correction for Healthy Sample",
     ylim = c(0,2.5),cex = 0.2,pch=20, xlim = c(0,6000))
abline(h=1,col=4, lwd = 1.5)
abline(v=4000,col=6,lt = 2, lwd = 1.2)
plot(fixed_2n_gc_med_eps[discontinuous_range], ylab = "After", xaxt = "n",
     ylim = c(0,2.5),cex = 0.2,pch=20, xlim = c(0,6000))
abline(h=1,col=4, lwd = 1.5)
abline(v=4000,col=6,lt = 2, lwd = 1.2)

plot(1)

```


```{r plot_gc_fix_healthy}
par(mfcol = c(2,1), mar = c(1,5,1,1))
# par(mfcol = c(2,1))
# Plot for Healthy
plot(fixed_2n_gc_med_eps[discontinuous_range], ylab = "Healthy", xaxt = "n",
     main = "After GC Correction ('One Sample')", 
     ylim = c(0,2), cex = 0.2, pch=20)
abline(h=1, col=4, lwd = 1.5)
abline(v=c(1000,1200,4000),col=c(5,5,6), lt=c(1,1,2), lwd = 1.2)
# Plot for Tumor
plot(fixed_2t_gc_med_eps[discontinuous_range], ylab = "Tumor", xaxt = "n",
     ylim = c(0,2), cex = 0.2, pch=20)
abline(h=1, col=4, lwd = 1.5)
abline(v=c(1000,1200,4000),col=c(5,5,6), lt=c(1,1,2), lwd = 1.2)
plot(1)

```


### Two sample corrections ...
```{r two_samp}
two_sample_correct = function(tumor, normal) {
  two_correct = (tumor)/(normal)
  two_correct_aft_med =  two_correct/median(two_correct,na.rm=TRUE)
  return(two_correct_aft_med)    
}
```

### Two sample corrections w/ epsilon
```{r two_samp_e}
two_sample_correct_e = function(tumor, normal,eps) {
  two_correct = (tumor+eps)/(normal+eps)
  two_correct_aft_med = two_correct/median(two_correct,na.rm=TRUE)
  return(two_correct_aft_med)
}

```


```{r apply_two_samp}
two_samp_med_gc_eps = two_sample_correct_e(fixed_2t_gc_med_eps,fixed_2n_gc_med_eps,0.1)
two_samp_med_eps = two_sample_correct_e(unfixed_2t_gc_med_eps,unfixed_2n_gc_med_eps,0.1)
two_samp_med = two_sample_correct(unfixed_2t_gc_med_eps,unfixed_2n_gc_med_eps)
```



```{r plot_two_samp}
# par(mfcol = c(2,1),mar = c(0.5,5,2,2))
par(mfcol = c(2,1),mar = c(0.5,5,2,2))
loc2bin = function(v, bin_size){
  return((v*1000000)/bin_size)}

bins = loc2bin(c(25,35,75,80),2500)
discontinuous_range = c(bins[1]:bins[2], bins[3]:bins[4])
# discontinuous_range = c(10000:14000,15000:16000)


## one sample median + GC correction
plot(unfixed_2t_gc_med_eps[discontinuous_range], ylab = "Before", xaxt = "n",
     main = "GC Correction for Tumor Sample",
     ylim = c(0,2.5),cex = 0.2,pch=20, xlim = c(0,6000))
abline(h=1,col=4, lwd = 1.5)
abline(v=4000,col=6,lt = 2, lwd = 1.2)
plot(fixed_2t_gc_med_eps[discontinuous_range], ylab = "After", xaxt = "n",
     ylim = c(0,2.5),cex = 0.2,pch=20, xlim = c(0,6000))
abline(h=1,col=4, lwd = 1.5)
abline(v=4000,col=6,lt = 2, lwd = 1.2)



## two sample median + GC correction
plot(fixed_2t_gc_med_eps[discontinuous_range], ylab = "Med+GC OneSamp (Tumor)", xaxt = "n",
     ylim = c(0,2.5),cex = 0.2,pch=20, xlim = c(0,6000))
abline(h=1,col=4, lwd = 1.5)
abline(v=4000,col=6,lt = 2, lwd = 1.2)

plot(two_samp_med_gc_eps[discontinuous_range], ylab = "Med+GC+Epsilon TwoSamp (Tumor)", xaxt = "n",
     ylim = c(0,2.5),cex = 0.2,pch=20, xlim = c(0,6000))
abline(h=1,col=4, lwd = 1.5)
abline(v=4000,col=6,lt = 2, lwd = 1.2)

plot(1)

```


```{r plot_fixes}
# par(mfcol = c(2,1), mar = c(1,5,1,1))
# par(mfcol = c(2,1))
# Plot for Healthy
plot(fixed_2n_gc_med_eps[discontinuous_range], ylab = "Healthy", xaxt = "n",
     main = "After GC Correction ('One Sample')", 
     ylim = c(0,2), cex = 0.2, pch=20)
abline(h=1, col=4, lwd = 1.5)
abline(v=c(1000,1200,4000),col=c(5,5,6), lt=c(1,1,2), lwd = 1.2)
# Plot for Tumor
plot(fixed_2t_gc_med_eps[discontinuous_range], ylab = "Tumor", xaxt = "n",
     ylim = c(0,2), cex = 0.2, pch=20)
abline(h=1, col=4, lwd = 1.5)
abline(v=c(1000,1200,4000),col=c(5,5,6), lt=c(1,1,2), lwd = 1.2)



####
####

plot(two_samp_med_gc_eps[discontinuous_range], xaxt = "n", 
     xlab = "", main = "Three Fixes: Median, GC, Two-Sample w/ Epsilon",
     ylim = c(0,2),cex = 0.2,pch=20)
abline(h=1,col=4, lwd = 1.5)
abline(v=c(1000,1200,4000),col=c(5,5,6), lt=c(1,1,2), lwd = 1.2)

plot(two_samp_med_gc_eps[discontinuous_range], xaxt = "n", 
     xlab = "", main = "Three Fixes: Median, GC, Two-Sample w/ Epsilon",
     ylim = c(0,2),cex = 0.2,pch=20)
abline(h=1,col=4, lwd = 1.5)
abline(v=c(1000,1200,4000),col=c(5,5,6), lt=c(1,1,2), lwd = 1.2)

plot(two_samp_med_eps[discontinuous_range],xaxt = "n",
     xlab = "", main = "Median + Two-Sample w/ Epsilon",
     ylim = c(0,2),cex = 0.2,pch=20)
abline(h=1,col=4, lwd = 1.5)
abline(v=c(1000,1200,4000),col=c(5,5,6), lt=c(1,1,2), lwd = 1.2)

plot(two_samp_med_eps[discontinuous_range],xaxt = "n",
     xlab = "Both Samples, Reads Count per Bin Location (2.5K)",
     main = "Median + Two-Sample (No Epsilon)",
     ylim = c(0,2),cex = 0.2,pch=20)
abline(h=1,col=4, lwd = 1.5)
abline(v=c(1000,1200,4000),col=c(5,5,6), lt=c(1,1,2), lwd = 1.2)
plot(1)

```

```{r plot_two_samp_fixes}
par(mfcol = c(3,1),mar = c(5,2,5,2))
plot(two_samp_med_gc_eps[discontinuous_range], xaxt = "n", 
     xlab = "", main = "Three Fixes: Median, GC, Two-Sample w/ Epsilon",
     ylim = c(0,2),cex = 0.2,pch=20)
abline(h=1,col=4, lwd = 1.5)
abline(v=c(1000,1200,4000),col=c(5,5,6), lt=c(1,1,2), lwd = 1.2)

plot(two_samp_med_eps[discontinuous_range],xaxt = "n",
     xlab = "", main = "Median + Two-Sample w/ Epsilon",
     ylim = c(0,2),cex = 0.2,pch=20)
abline(h=1,col=4, lwd = 1.5)
abline(v=c(1000,1200,4000),col=c(5,5,6), lt=c(1,1,2), lwd = 1.2)

plot(two_samp_med_eps[discontinuous_range],xaxt = "n",
     xlab = "Both Samples, Reads Count per Bin Location (2.5K)",
     main = "Median + Two-Sample (No Epsilon)",
     ylim = c(0,2),cex = 0.2,pch=20)
abline(h=1,col=4, lwd = 1.5)
abline(v=c(1000,1200,4000),col=c(5,5,6), lt=c(1,1,2), lwd = 1.2)

plot(1)
```




