---
title: "Lecture3_Copy_v2"
author: "Yuval Benjamini (Carmel Edit)"
date: Sys.Date()
output:
  html_document:
    code_folding: show
  pdf_document: default
editor_options:
  markdown:
    wrap: sentence
---

```{r}
library('data.table')
library('tictoc')
library('dplyr')
source("help_funcs.R")
```

Load data...
```{r}

# Benjamini:
# reads_file = 'C:/Users/user/Documents/Carmel/Projects/HUJI/Archive_HUJI/data-analysis-genomics/data/TCGA-13-0723-10B_lib2_all_chr1.forward'

# Group A:
reads_file = 'C:/Users/user/Documents/Carmel/Projects/HUJI/Archive_HUJI/data-analysis-genomics/data/TCGA-13-0723-01A_lib2_all_chr1.forward'

chr1_reads_orig = fread(reads_file) 
colnames(chr1_reads_orig) = c("Chrom","Loc","FragLen")   


```

You should have a file of "helper functions" that you call using `source("functions.R")`. 

# Uniform Poisson model

This week we explored the distribution of DNA-sequenced fragments. \n
By definition, locations accept only integers and therefore behave like a discrete random variable. \n 
In particular, locations along a given chromosome sequence behave like a random variable from a Poisson distribution.

First, let us produce many instances of uniform variables:
```{r}
set.seed(1)
nsamp = 100e+06 # 100M
max_samp = nsamp * 10
unif_dat = runif(n = nsamp, 0,max_samp )
s_dat = sort(unif_dat)
```

Let's take a look at the first few locations 

```{r}
plot(s_dat[1:1000],rep(1,1000), xlim = c(1,1000))
```

The intervals between uniform variables behave like an exponential variable:

```{r}
diffs = s_dat[2:nsamp] - s_dat[(1:nsamp-1)]
hist(diffs,freq = FALSE)
curve(dexp(x,rate = 1/10),add = TRUE)
```

Here the blue curve is the exponential distribution.
In an exponential process the increments are independent.


We get the Poisson by counting occurrences in equal-size intervals,
as defined by the bin size (denoted $k$). \n
Since the intervals are of equal size, we expect to have the same probability to end up in any of the bins. \n
In other words, we expect to have an average of one occurrence per bin.


```{r}
k = 1000
output_1K = binReads(ceiling(s_dat), k)
range(output_1K) #includes NA values
out_clean = range(output_1K,na.rm = TRUE) #excludes NA
out_clean
min_1K = out_clean[1] - 2
max_1K = out_clean[2]
hist(output_1K) #includes NA values (9e+05) #frequency
hist(output_1K,seq(min_1K,max_1K,4),freq = TRUE) #frequency
hist(output_1K,seq(min_1K,max_1K,4),freq = FALSE) #density

```

So this still looks like a Poisson: 

```{r}
range(output_1K)
range(output_1K,na.rm = TRUE)
hist(output_1K,seq(61,149,4),freq = FALSE)
pois_probs = dpois(x = 60:147, lambda = 1000/10)
pois_probs_4s = rowMeans(matrix(pois_probs,byrow = TRUE,nc = 4))
points(seq(62,147,4),pois_probs_4s)
```


## Comparing the Poisson model to coverage

```{r}

# get read line of actual data
beg_region = 1
end_region = 10e06
locations = chr1_reads_orig$Loc
N = end_region-beg_region+1


brks=40
read_line = getReadLine(locations, beg_region, end_region)
read_hist = hist(read_line,breaks = 0:brks-0.5)
text(read_hist$mids,read_hist$counts,read_hist$counts,
     pos = 3,cex =0.5)

```


Is the distribution of reads uniform? 
```{r}
reg_lambda=sum(read_line)/length(read_line)     # in region
dpois(0:brks,lambda = reg_lambda) # very hard to see the probabilities

bb=brks-1
obs_prob = read_hist$density
model_prob = dpois(0:bb,lambda = reg_lambda)
plot(obs_prob, model_prob)
abline(0,1)

plot(log(obs_prob), log(model_prob))
text(log(obs_prob), log(model_prob),0:bb,pos=2, cex=0.7)
abline(0,1)
# Or
plot(log(obs_prob), log(model_prob),xlim = c(-90,0),ylim = c(-90,0))
text(log(obs_prob), log(model_prob),0:bb,pos=2, cex=0.7)
abline(0,1)
```


## Preparing a coverage vector for bins of 5K...
```{r}

read_line = getReadLine(locations, beg_region, end_region)
reads_20K = binReads(input = read_line,size = 20e+03)

brks=40
read_hist = hist(read_line,breaks = 0:brks-0.5)
text(read_hist$mids,read_hist$counts,read_hist$counts,
     pos = 3,cex =0.5)
obs_prob = read_hist$density
pois_probs = dpois(0:660,mean(reads_20K))


```


## Comparison to poisson.
```{r}
# compute poisson probabilities and compare to observed probabilities...

# Compare these to the probabilities from the bulk. 
# What does this model imply for the probability of being between X_i in 100:109?
sd(reads_20K,na.rm=TRUE)

```



