---
title: "Lab Task 8"
author: "Group 6"
date: "`r format(Sys.time(), '%d/%m/%y')`"
output:
  html_document:
    code_folding: show
editor_options: 
  markdown: 
    wrap: sentence
---

Group members:

- Shawn Yakir , 315714980,  shawn.yakir@mail.huji.ac.il
- Sarah Levitz, 324673623, sarah.levitz@mail.huji.ac.il
- Carmel Baris, 318455276, carmel.baris@mail.huji.ac.il 

### Clean the environment
```{r clean}
rm(list = ls())
```

### Libraries used

```{r libraries}
library('splines')
library('scales')
```


### Paths
```{r paths}
path = "C:/Users/user/Documents/Carmel/Projects/HUJI/Archive_HUJI/data-analysis-genomics/data"
# path = "/Users/alevi/Downloads/"
 # path = "\\Users\\shawn\\Desktop\\"
```


### Functions

Correcting functions:

```{r correcting_funcs}

### Median correction
fix_med = function(samp) {
  return(samp/median(samp,na.rm=TRUE))
}

### One sample correction
fix_gc_e = function(samp, preds, eps=0) {
  return((samp+eps)/(preds+eps))
}


### Two sample correction
fix_2samp_e = function(tumor, healthy, eps=0) {
  return((tumor+eps)/(healthy+eps))
}

```


Fitting bspline regression:

```{r fit}
f = function(X, Y, deg=3, knots=c(0.415, 0.45, 0.54), subs) {
  lm(Y ~ bs(X, degree = deg, knots = knots, Boundary.knots = c(min(X), max(X))),
     subset = subs)
}

```


### Load data
```{r dat}
load(sprintf("%s/reads_100_A2.rda",path))
load(sprintf("%s/reads_100_B2.rda",path))
load(sprintf("%s/GC_100.rda",path))
```

### Binning
```{r binning}
binsize = 25

chr2t_mat = matrix(reads_100_A2, nr = binsize)
chr2n_mat = matrix(reads_100_B2, nr = binsize)
GC_100=GC_100[1: (length(reads_100_A2)) ]
gc_mat = matrix(GC_100,nr=binsize)

reads2.5K_t2 = colSums(chr2t_mat)
reads2.5K_n2 = colSums(chr2n_mat)
GC2.5K = colMeans(gc_mat)

length(reads2.5K_n2) == length(reads2.5K_t2)

## ------------------------------------------------
rm("chr2n_mat","chr2t_mat","gc_mat","reads_100_A2","reads_100_B2","GC_100") # remove from memory
```


### Outliers
```{r outliers_new}

find_out_boxplot <- function(dat) {
  out = boxplot.stats(dat)$out
  out_ind = dat %in% out
  return(as.integer(out_ind))
}

# zeros_GC = GC2.5K == 0 #bins w/o GC content
# small_GC = GC2.5K <= 0.2 & GC2.5K > 0 #low-GC content bins (<30%)

out_t2 = find_out_boxplot(reads2.5K_t2)
out_n2 = find_out_boxplot(reads2.5K_n2)
out_reads_xtra = (reads2.5K_t2<5) | (reads2.5K_n2<5)


outliers = (out_t2 | out_n2 | out_reads_xtra)
r_health_cln = reads2.5K_n2[!outliers]
r_tumor_cln = reads2.5K_t2[!outliers]
GC_cln = GC2.5K[!outliers]

cat("removed", round((1-(length(GC_cln)/length(GC2.5K)))*100,2),"% outliers")


# identify outliers of reads count per bin (0.2%)
# z_scores_n2 = (reads2.5K_n2 - mean(reads2.5K_n2)) / sd(reads2.5K_n2)
# z_scores_t2 = (reads2.5K_t2 - mean(reads2.5K_t2)) / sd(reads2.5K_t2)
# xtreme_reads_n2 = abs(z_scores_n2) > 2
# xtreme_reads_t2 = abs(z_scores_t2) > 2
# outliers_alt = (xtreme_reads_n2 | xtreme_reads_t2)
# r_health_cln = reads2.5K_n2[!outliers_alt]
# r_tumor_cln = reads2.5K_t2[!outliers_alt]
# GC_cln = GC2.5K[!outliers_alt]
# cat(round((1-(length(GC_cln)/length(GC2.5K)))*100,2),"% removed")



## ------------------------------------------------
rm("path", "find_out_boxplot","out_n2","out_reads_xtra","out_t2") # remove from memory
```


### Healthy vs Tumor Reads (distribution against GC content)
```{r n2t2_gc}
set.seed(50)
samp1 = sample(length(r_health_cln), 10000)

par(mfcol = c(2, 2), mar = c(1.1, 5, 5, 1))

plot(GC_cln[samp1], r_health_cln[samp1], 
     cex = 0.4, pch = 20, col = rgb(0, 0, 0, 0.2),
     ylim = c(0, 400), xlim = c(0.2, 0.75), 
     ylab = "Healthy 2", main = "Original")
plot(GC_cln[samp1], r_tumor_cln[samp1], 
     cex = 0.4, pch = 20, col = rgb(0, 0, 0, 0.2),
     ylim = c(0, 400), xlim = c(0.2, 0.75),
     ylab = "Tumor 2",)

### Normalized for Comparison
# Use the rescale function from the scales package
# r_n2_cl_scaled = scales::rescale(r_health_cln, to = c(0, 1000))
# r_t2_cl_scaled = scales::rescale(r_tumor_cln, to = c(0, 1000))

# Plot the normalized data
plot(GC_cln[samp1], scales::rescale(r_health_cln, to = c(0, 1000))[samp1], 
     cex = 0.4, pch = 20, col = rgb(0, 0, 0, 0.2),
     ylim = c(0, 1000), xlim = c(0.2, 0.75), 
     xaxt = "n",  ylab = "", main = "Scaled")
plot(GC_cln[samp1], scales::rescale(r_tumor_cln, to = c(0, 1000))[samp1], 
     cex = 0.4, pch = 20, col = rgb(0, 0, 0, 0.2),
     ylim = c(0, 1000), xlim = c(0.2, 0.75), 
     xaxt = "n", ylab = "")
plot(1)
```


### Learn parameters from chromosome 1

```{r}
l_2n = length(r_health_cln)
l_2t = length(r_tumor_cln)

myknots = c(0.415, 0.45, 0.54)
pred_range = seq(0.2,0.75,0.01)

GC=GC_cln[1:l_2n]
r_health_cln = r_health_cln
r_tumor_cln = r_tumor_cln

subs_n = (r_health_cln>50)&(r_health_cln<350)
subs_t = (r_tumor_cln>50)&(r_tumor_cln<350)

# mod_2n = f(GC, r_health_cln, subs = subs_n)
# mod_2t = f(GC, r_tumor_cln, subs = subs_n)
mod_2n = lm(r_health_cln~bs(GC, degree = 3),subset = subs_n)
mod_2t = lm(r_tumor_cln~bs(GC, degree = 3),subset = subs_t)

resp2n = predict(mod_2n,list(GC=pred_range))
preds2n = predict(mod_2n,list(GC=GC_cln))

resp2t = predict(mod_2t,list(GC=pred_range))
preds2t = predict(mod_2t,list(GC=GC_cln))

par(mfcol = c(1,2),mar = c(3,5,3,3))

plot(GC_cln[samp1],r_health_cln[samp1],
     cex = 0.5, pch=20, col = rgb(0, 0, 0, 0.2),
     ylim = c(0,350),xlim = c(0.2,.75), 
     ylab = "Cleaned Reads (bins of 2,500)", main = "Healthy 2")
lines(pred_range, resp2n,col = 4,lw=3)

plot(GC_cln[samp1],r_tumor_cln[samp1],
     cex = 0.5, pch=20, col = rgb(0, 0, 0, 0.2),
     ylim = c(0,350),xlim = c(.2,.75), 
     ylab = "", main = "Tumor 2")
lines(pred_range, resp2t,col = 2,lw=3)
lines(pred_range, resp2n,col = 4,lw=3)

plot(1)
```


### Healthy Fit vs. Tumor Fit

```{r}
# r_tumor_cln = r_tumor_cln[1:l_2n]  #trim tumor data to length of healthy

# subs_n = (r_health_cln>50)&(r_health_cln<350)
# subs_t = (r_tumor_cln>50)&(r_tumor_cln<350)

# mod_2n = f(GC, r_health_cln, subs = subs_n)
# mod_2t = f(GC, r_tumor_cln, subs = subs_n)

# Fit B-spline Regression Models
mod_healthy = lm(r_health_cln ~ bs(GC_cln, degree = 3))
mod_tumor = lm(r_tumor_cln ~ bs(GC_cln, degree = 3))


preds_healthy = predict(mod_healthy, newdata = data.frame(GC_cln = pred_range))
preds_tumor = predict(mod_tumor, newdata = data.frame(GC_cln = pred_range))

```


```{r}
# Combined Plot
par(mfcol = c(1, 1), mar = c(5, 5, 4, 2))

# Plot the data points first
plot(GC_cln, r_tumor_cln, col = rgb(0.5, 0.5, 0.5, 0.5), pch = 20, cex = 0.5, ylab = "Reads", xlab = "GC Content", main = "Healthy vs Tumor Sample", ylim = c(0, 350))
points(GC_cln, r_health_cln, col = rgb(0.8, 0.8, 0.8, 0.5), pch = 20, cex = 0.5)

# Add the regression lines on top
lines(pred_range, preds_healthy, col = 'darkblue', lwd = 2)
lines(pred_range, preds_tumor, col = 'darkred', lwd = 2)

# Add legend
legend("topleft", legend = c("Tumor Sample", "Healthy Sample"), col = c("gray35", "lightgray"), pch = 20, bty = "n", cex = 1.2)

```


### Spatial Distribution of Normative vs Tumor Reads
```{r}

par(mfcol = c(2, 1), mar = c(1, 5, 1, 1))
plot(r_health_cln[samp1], 
     cex = 0.4, pch = 20, col = rgb(0,0,0,0.2),
     xaxt = "n", ylab = "Healthy 2", ylim = c(0, 500))
plot(r_tumor_cln[samp1], 
     cex = 0.4, pch = 20, col = rgb(0,0,0,0.2),
     xaxt = "n", ylab = "Tumor 2", ylim = c(0, 500))

# par(mfcol =c(2,1),mar =c(1,1,1,1))
# set.seed(50)
# samp1 = sample(length(r_health_cln),50000)

# plot(GC2.5K[samp1],r_health_cln[samp1],ylim = c(0, 350),xlim = c(0.2,.75), cex = 0.5, pch=20)
# plot(GC2.5K[samp1],r_tumor_cln[samp1],ylim = c(0, 350),xlim = c(0.2,.75),cex = 0.5, pch=20)

```



### Learn parameters from chromosome 1

```{r}
l_2n = length(r_health_cln)
l_2t = length(r_tumor_cln)
GC=GC_cln[1:l_2n]
length(GC)

```



### One sample corrections (aka only GC & median correction)
```{r}
one_sample_gc_correct = function(samp, preds, eps) {
  one_correct = (samp+eps)/(preds+eps)
  adj_med = 1/median(one_correct)
  return(adj_med * one_correct)
}
```

```{r}
# take maximum with 0
preds1n = pmax(predict(mod_2n,list(GC=GC_cln)),0)
preds1t = pmax(predict(mod_2t,list(GC=GC_cln)),0)


# One Sample with espilon, Normative
eps = 0.1
fixed_2n_gc_med_eps = fix_med(fix_gc_e(r_health_cln, preds1n[1:l_2n], eps))
fixed_2n_gc_med_eps0 = one_sample_gc_correct(r_health_cln, preds1n[1:l_2n], eps)

unfixed_2n_gc_med_eps = fix_med(fix_gc_e(r_health_cln, rep(1,l_2n), eps))
unfixed_2n_gc_med_eps0 = one_sample_gc_correct(r_health_cln, rep(1,l_2n), eps)

# One Sample with espilon, Tumor
eps = 0.1
fixed_2t_gc_med_eps = fix_med(fix_gc_e(r_tumor_cln, preds1t[1:l_2t], eps))
fixed_2t_gc_med_eps0 = one_sample_gc_correct (r_tumor_cln, preds1t[1:l_2t], eps)

unfixed_2t_gc_med_eps = fix_med(fix_gc_e(r_tumor_cln, rep(1,l_2t), eps))
unfixed_2t_gc_med_eps0 = one_sample_gc_correct(r_tumor_cln, rep(1,l_2t), eps)
```


```{r}
par(mfcol = c(2,1),mar = c(2,5,2,2))
loc2bin = function(v, bin_size){
  return((v*1000000)/bin_size)}

bins = loc2bin(c(25,35,75,80),2500)
stiched_range = c(bins[1]:bins[2], bins[3]:bins[4])
# stiched_range = c(10000:14000,15000:16000)
plot(unfixed_2t_gc_med_eps[stiched_range], 
     main = "GC Fix for Tumor Sample",
     ylab = "Before", xaxt = "n",
     ylim = c(0,2),cex = 0.2,pch=20, xlim = c(0,6000))
abline(h=1,col=4, lwd = 1.5)
abline(v=4000,col=6,lt = 2, lwd = 1.2)

plot(fixed_2t_gc_med_eps[stiched_range], 
     ylab = "After", xaxt = "n",
     ylim = c(0,2),cex = 0.2,pch=20, xlim = c(0,6000))
abline(h=1,col=4, lwd = 1.5)
abline(v=4000,col=6,lt = 2, lwd = 1.2)
mtext("Selected Locations in Chrom1 (25M-35M, 75M-80M)", side = 1, cex = 0.8)

# plot(1)
## ------------------------------------------------

plot(unfixed_2n_gc_med_eps[stiched_range], ylab = "Before", xaxt = "n",
     main = "GC Correction for Healthy Sample",
     ylim = c(0,2),cex = 0.2,pch=20, xlim = c(0,6000))
abline(h=1,col=4, lwd = 1.5)
abline(v=4000,col=6,lt = 2, lwd = 1.2)
plot(fixed_2n_gc_med_eps[stiched_range], ylab = "After", xaxt = "n",
     ylim = c(0,2),cex = 0.2,pch=20, xlim = c(0,6000))
abline(h=1,col=4, lwd = 1.5)
abline(v=4000,col=6,lt = 2, lwd = 1.2)
mtext("Selected Locations in Chrom1 (25M-35M, 75M-80M)", side = 1, cex = 0.8)


# plot(1)

## ------------------------------------------------

# Plot for Healthy
plot(fixed_2n_gc_med_eps[stiched_range], 
     ylab = "Healthy",xaxt = "n",
     main = "After 'One Sample' GC Fix",
     ylim = c(0,2), cex = 0.2, pch=20)
abline(h=1, col=4, lwd = 1.5)
abline(v=c(750,900,4000),col=c(5,5,6), lt=c(1,1,2), lwd = 1.2)
# Plot for Tumor
plot(fixed_2t_gc_med_eps[stiched_range], 
     ylab = "Tumor", xaxt = "n",
     ylim = c(0,2), cex = 0.2, pch=20)
abline(h=1, col=4, lwd = 1.5)
abline(v=c(750,900,4000),col=c(5,5,6), lt=c(1,1,2), lwd = 1.2)
mtext("Selected Locations in Chrom1 (25M-35M, 75M-80M)", side = 1, cex = 0.8)

plot(1)

```


### Two sample corrections ...
```{r}
two_sample_correct = function(tumor, normal) {
  two_correct = (tumor)/(normal)
  two_correct_aft_med =  two_correct/median(two_correct,na.rm=TRUE)
  return(two_correct_aft_med)    
}
```

### Two sample corrections w/ epsilon
```{r}
two_sample_correct_e = function(tumor, normal, eps) {
  two_correct = (tumor+eps)/(normal+eps)
  two_correct_aft_med = two_correct/median(two_correct,na.rm=TRUE)
  return(two_correct_aft_med)
}

```


```{r}
two_samp_med_gc_eps = two_sample_correct_e(fixed_2t_gc_med_eps,fixed_2n_gc_med_eps,0.1)
two_samp_med_gc_eps0 = two_sample_correct_e(fixed_2t_gc_med_eps,fixed_2n_gc_med_eps,0.1)

two_samp_med_eps = two_sample_correct_e(unfixed_2t_gc_med_eps,unfixed_2n_gc_med_eps,0.1)
two_samp_med_eps0 = two_sample_correct_e(unfixed_2t_gc_med_eps,unfixed_2n_gc_med_eps,0.1)

two_samp_med = two_sample_correct(unfixed_2t_gc_med_eps,unfixed_2n_gc_med_eps)
two_samp_med0 = two_sample_correct(unfixed_2t_gc_med_eps,unfixed_2n_gc_med_eps)
```



```{r}
par(mfcol = c(2,1),mar = c(0.5,5,2,2))

## two sample median + GC correction
plot(fixed_2t_gc_med_eps[stiched_range], 
     main = "All Corrections for Tumor Sample", sub = "Med + GC (OneSamp)",
     ylab = "One Sample", xaxt = "n",
     ylim = c(0,2),cex = 0.2,pch=20, xlim = c(0,6000))
abline(h=1,col=4, lwd = 1.5)
abline(v=4000,col=6,lt = 2, lwd = 1.2)

plot(two_samp_med_gc_eps[stiched_range], sub = "Med + GC + TwoSamp",
     ylab = "Two Sample", xaxt = "n",
     ylim = c(0,2),cex = 0.2,pch=20, xlim = c(0,6000))
abline(h=1,col=4, lwd = 1.5)
abline(v=4000,col=6,lt = 2, lwd = 1.2)

plot(1)

```


```{r}
# par(mfcol = c(2,1), mar = c(1,5,1,1))
# par(mfcol = c(2,1))
# Plot for Healthy
plot(fixed_2n_gc_med_eps[stiched_range], ylab = "Healthy", xaxt = "n",
     main = "After GC Correction ('One Sample')", 
     ylim = c(0,2), cex = 0.2, pch=20)
abline(h=1, col=4, lwd = 1.5)
abline(v=c(1000,1200,4000),col=c(5,5,6), lt=c(1,1,2), lwd = 1.2)
# Plot for Tumor
plot(fixed_2t_gc_med_eps[stiched_range], ylab = "Tumor", xaxt = "n",
     ylim = c(0,2), cex = 0.2, pch=20)
abline(h=1, col=4, lwd = 1.5)
abline(v=c(1000,1200,4000),col=c(5,5,6), lt=c(1,1,2), lwd = 1.2)



####
####

plot(two_samp_med_gc_eps[stiched_range], xaxt = "n", 
     xlab = "", main = "Three Fixes: Median, GC, Two-Sample w/ Epsilon",
     ylim = c(0,2),cex = 0.2,pch=20)
abline(h=1,col=4, lwd = 1.5)
abline(v=c(1000,1200,4000),col=c(5,5,6), lt=c(1,1,2), lwd = 1.2)

plot(two_samp_med_eps[stiched_range],xaxt = "n",
     xlab = "", main = "Median + Two-Sample w/ Epsilon",
     ylim = c(0,2),cex = 0.2,pch=20)
abline(h=1,col=4, lwd = 1.5)
abline(v=c(1000,1200,4000),col=c(5,5,6), lt=c(1,1,2), lwd = 1.2)

plot(two_samp_med_eps[stiched_range],xaxt = "n",
     xlab = "Both Samples, Reads Count per Bin Location (2.5K)",
     main = "Median + Two-Sample (No Epsilon)",
     ylim = c(0,2),cex = 0.2,pch=20)
abline(h=1,col=4, lwd = 1.5)
abline(v=c(1000,1200,4000),col=c(5,5,6), lt=c(1,1,2), lwd = 1.2)
plot(1)

```

```{r}
par(mfcol = c(3,1),mar = c(5,2,5,2))
plot(two_samp_med_gc_eps[stiched_range], xaxt = "n", 
     xlab = "", main = "Three Fixes: Median, GC, Two-Sample w/ Epsilon",
     ylim = c(0,2),cex = 0.2,pch=20)
abline(h=1,col=4, lwd = 1.5)
abline(v=c(1000,1200,4000),col=c(5,5,6), lt=c(1,1,2), lwd = 1.2)

plot(two_samp_med_eps[stiched_range],xaxt = "n",
     xlab = "", main = "Median + Two-Sample w/ Epsilon",
     ylim = c(0,2),cex = 0.2,pch=20)
abline(h=1,col=4, lwd = 1.5)
abline(v=c(1000,1200,4000),col=c(5,5,6), lt=c(1,1,2), lwd = 1.2)

plot(two_samp_med_eps[stiched_range],xaxt = "n",
     xlab = "Both Samples, Reads Count per Bin Location (2.5K)",
     main = "Median + Two-Sample (No Epsilon)",
     ylim = c(0,2),cex = 0.2,pch=20)
abline(h=1,col=4, lwd = 1.5)
abline(v=c(1000,1200,4000),col=c(5,5,6), lt=c(1,1,2), lwd = 1.2)

plot(1)
```




